### PR TITLE
Adding functional batch_norm to BatchNorm2d substitution

### PR DESCRIPTION
**Pull request by edenlum**
## Checklist before requesting a review:
- [x] I set the appropriate labels on the pull request.
- [x] I have added/updated the release note draft (if necessary).
- [x] I have updated the documentation to reflect my changes (if necessary).
- [x] All function and files are well documented. 
- [x] All function and classes have type hints.
- [x] There is a licenses in all file.
- [x] The function and variable names are informative. 
- [x] I have checked for code duplications.
- [x] I have added new unittest (if necessary).

### CODE REVIEW

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_batch_norm.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
+# Copyright 2022 Sony Semiconductor Israel, Inc. All rights reserved.
```
Comment (haihabi) [inline model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_batch_norm.py:1]: We are in 2023

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_batch_norm.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+            beta = list(input_nodes[2].weights.values())[0]
+            mean = list(input_nodes[3].weights.values())[0]
+            var = list(input_nodes[4].weights.values())[0]
+        elif len(input_nodes) == 3:
+            mean = list(input_nodes[1].weights.values())[0]
+            var = list(input_nodes[2].weights.values())[0]
+            gamma = None
+            beta = None
+        else:
+            raise ValueError(f'functional batch_norm is expected to have 3 or 5 inputs, got {len(input_nodes)}')
```
Comment (haihabi) [inline model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_batch_norm.py:62]: Use logger

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_batch_norm.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+                                                   MOMENTUM: MOMENTUM_VAL},
+                                   input_shape=node.output_shape,
+                                   output_shape=node.output_shape,
+                                   weights=bn_node_weights,
+                                   layer_class=nn.BatchNorm2d)
+
+        num_nodes_before_substitution = len(graph.nodes)
+        num_edges_before_substitution = len(graph.edges)
+
+        # new_batchnorm2d.prior_info = copy.deepcopy(node.prior_info)
```
Comment (haihabi) [inline model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_batch_norm.py:96]: Remove comment

**Comment by haihabi** [comment]
Also link this PR to the open issue

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_batch_norm.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+import torch.nn.functional as F
+import copy
+
+from model_compression_toolkit.core.common.graph.graph_matchers import NodeOperationMatcher
+from model_compression_toolkit.core import common
+from model_compression_toolkit.core.common import BaseNode, Graph, FrameworkInfo
+from model_compression_toolkit.core.pytorch.constants import *
+
+
+def functional_batch_norm_matcher() -> NodeOperationMatcher:
```
Comment (elad-c) [inline model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_batch_norm.py:25]: Where is this function used?

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_batch_norm.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    bn_node = NodeOperationMatcher(F.batch_norm)
+    return bn_node
+
+
+class FunctionalBatchNorm(common.BaseSubstitution):
+    """
+    Replace functional batch_norm with BatchNorm2d.
+    """
+
+    def __init__(self, ):
```
Comment (elad-c) [inline model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_batch_norm.py:42]: remove ,

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_batch_norm.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        """
+        Matches: functional batch_norm
+        """
+        bn_node = NodeOperationMatcher(F.batch_norm)
+        super().__init__(matcher_instance=bn_node)
+
+    def get_attributes_from_inputs(self, graph: Graph, node: BaseNode) -> dict:
+        input_nodes = graph.get_prev_nodes(node)
+        if len(input_nodes) == 5:
```
Comment (elad-c) [inline model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_batch_norm.py:51]: make sure len(input_nodes) == 4 isn't an option

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_batch_norm.py [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+                MOVING_VARIANCE: var}
+
+    def substitute(self,
+                   graph: Graph,
+                   node: BaseNode) -> Graph:
+        """
+        Substitute functional.batch_norm and its inputs with BatchNorm2d.
+        Args:
+            graph: Graph we apply the substitution on.
+            node: nodes that match the pattern in the substitution init.
```
Comment (elad-c) [inline model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_batch_norm.py:76]: single "node"

File: tests/pytorch_tests/model_tests/feature_models/bn_folding_test.py [hunk 8]
```diff
@@ -13,6 +13,7 @@
 # limitations under the License.
 # ==============================================================================
 import torch
+from torch import nn
```
Comment (elad-c) [inline tests/pytorch_tests/model_tests/feature_models/bn_folding_test.py:16]: can remove the torch import

File: tests/pytorch_tests/model_tests/test_feature_models_runner.py [hunk 9]
```diff
@@ -131,12 +137,21 @@ def test_bn_folding(self):
...
-        BNFoldingNetTest(self, nn.Conv2d(3, 3, kernel_size=3, groups=3)).run_test()  # DW-Conv test
-        BNFoldingNetTest(self, nn.ConvTranspose2d(3, 2, kernel_size=(2, 1))).run_test()
-        BNFoldingNetTest(self, nn.Conv2d(3, 2, kernel_size=2), fold_applied=False).run_test()
-        BNFoldingNetTest(self, nn.Conv2d(3, 3, kernel_size=(3, 1), groups=3), fold_applied=False).run_test()  # DW-Conv test
-        BNFoldingNetTest(self, nn.ConvTranspose2d(3, 2, kernel_size=(1, 3)), fold_applied=False).run_test()
+        assert False
+
+        def batch_norm_wrapper(channels):
+            return partial(nn.functional.batch_norm,
+                           running_mean=torch.zeros(channels, device='cuda'),
```
Comment (elad-c) [inline tests/pytorch_tests/model_tests/test_feature_models_runner.py:144]: init with somthing other that _zeros_ and _ones_. maybe _random_

File: tests/pytorch_tests/model_tests/test_feature_models_runner.py [hunk 10]
```diff
@@ -131,12 +137,21 @@ def test_bn_folding(self):
...
         """
-        BNFoldingNetTest(self, nn.Conv2d(3, 2, kernel_size=1)).run_test()
-        BNFoldingNetTest(self, nn.Conv2d(3, 3, kernel_size=3, groups=3)).run_test()  # DW-Conv test
-        BNFoldingNetTest(self, nn.ConvTranspose2d(3, 2, kernel_size=(2, 1))).run_test()
-        BNFoldingNetTest(self, nn.Conv2d(3, 2, kernel_size=2), fold_applied=False).run_test()
-        BNFoldingNetTest(self, nn.Conv2d(3, 3, kernel_size=(3, 1), groups=3), fold_applied=False).run_test()  # DW-Conv test
-        BNFoldingNetTest(self, nn.ConvTranspose2d(3, 2, kernel_size=(1, 3)), fold_applied=False).run_test()
+        assert False
+
+        def batch_norm_wrapper(channels):
```
Comment (elad-c) [inline tests/pytorch_tests/model_tests/test_feature_models_runner.py:142]: Add wrapper for F.batch_norm with _weight_ and _bias_

**Review by elad-c — CHANGES_REQUESTED** [review]

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_batch_norm.py (hunk 2 above)
Comment (edenlum) [reply model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_batch_norm.py:62]: use Logger.error(msg), like in gptq_training.py. is that the correct function?

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_batch_norm.py (hunk 4 above)
Comment (edenlum) [reply model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_batch_norm.py:25]: deleted

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_batch_norm.py (hunk 6 above)
Comment (edenlum) [reply model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_batch_norm.py:51]: there's and error if "else"

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_batch_norm.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+            mean = list(input_nodes[3].weights.values())[0]
+            var = list(input_nodes[4].weights.values())[0]
+        elif len(input_nodes) == 3:
+            mean = list(input_nodes[1].weights.values())[0]
+            var = list(input_nodes[2].weights.values())[0]
+            gamma = None
+            beta = None
+        else:
+            Logger.error(f'functional batch_norm is expected to have 3 or 5 inputs, got {len(input_nodes)}')
+
```
Comment (Idan-BenAmi) [inline model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_batch_norm.py:51]: Consider to check the tensor dimensions (as you only support banchNorm2d)

Comment (edenlum) [reply model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_batch_norm.py:48]: done

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_batch_norm.py (hunk 6 above)
Comment (elad-c) [reply model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_batch_norm.py:40]: I saw that. I meant that it's not an option in torch. for example if it is craeted like this:
`torch.nn.functional.batch_norm(input, running_mean, running_var, weight=weight, bias=None`
the bias is still created

File: tests/pytorch_tests/model_tests/test_feature_models_runner.py [hunk 12]
```diff
@@ -137,42 +137,54 @@ def test_bn_folding(self):
         """
         This test checks the BatchNorm folding feature.
         """
-        assert False
 
         def batch_norm_wrapper(channels):
             return partial(nn.functional.batch_norm,
-                           running_mean=torch.zeros(channels, device='cuda'),
-                           running_var=torch.ones(channels, device='cuda'))
+                           running_mean=0+torch.randn(channels, device='cuda'),
```
Comment (elad-c) [inline tests/pytorch_tests/model_tests/test_feature_models_runner.py:143]: you'll probably need to select the device accordingly, because tests are ran on CPU in github

**Review by elad-c — COMMENTED** [review]
See leftover comments

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_batch_norm.py (hunk 6 above)
Comment (edenlum) [reply model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_batch_norm.py:40]: if weight\bias = None is doesn't create the input node

**Review by Idan-BenAmi — APPROVED** [review]

**Review by elad-c — APPROVED** [review]
